Load lldp and enabled/disabled status on Kytos bootstrap

cmagnobarbosa opened the pull request:
Today, to restore administrative changes, it is necessary to use REST Endpoint `restore`. This PR changes the `restore_network_status` method responsible for this task, making this task automatic in the kytos bootstrap.

Main changes:
- The state stored in the storehouse is loaded into NApp `execute`.
- The states are restored when the switch or link is discovered by the topology.
- Remove the REST Endpoint `\restore`.
- In the kytos bootstrap, the state saved in the storehouse is restored.

**Note**: To use this change, it is recommended to clear the` kytos.topology.status` namespace in storehouse.

Main updates:
- Update unit tests.
- Update the storehouse client to expect to recover or create a box in the storehouse.

Fix #124

cmagnobarbosa committed:
The process for restoring administrative changes is now automatic

Today, to restore administrative changes it is necessary to use REST Endpoint
`restore`. This commit changes the restore_network_status method, making this
task automatic in the kytos bootstrap.

cmagnobarbosa committed:
Add a timer to wait before restoring the administrative state of the interfaces

This commit adds a timer to wait for switch.interfaces to be filled.
This avoids a false state of success in restoring the entire administrative
state of a switch's interfaces. Also, small refactoring was done in the method
 `_restore_switches` allowing these modifications.

ajoaoff commented on main.py:
```diff
@@ -190,16 +197,17 @@ def get_topology(self):
+            elif isinstance(obj, Link):
+                if obj.id not in self.links_verified:
+                    self.links_verified.append(obj.id)
+                    self._restore_links(obj.id)
```
```suggestion
                    self._restore_link(obj.id)
```

ajoaoff commented on main.py:
```diff
@@ -190,16 +197,17 @@ def get_topology(self):
-            self._load_network_status()
-            self._restore_status()
+            if isinstance(obj, Switch):
+                self._restore_switches(obj.id)
```
```suggestion
                self._restore_switch(obj.id)
```

ajoaoff commented on main.py:
```diff
@@ -91,72 +96,75 @@ def _get_link_from_interface(self, interface):
+            raise KeyError(error)
+        log.info(f'The state of link {link_id} has been restored.')
+
+    def _restore_switches(self, switch_id):
```
Same here.
```suggestion
    def _restore_switch(self, switch_id):
```

ajoaoff commented on main.py:
```diff
@@ -91,72 +96,75 @@ def _get_link_from_interface(self, interface):
-            endpoint_b.nni = True
-
-            self.notify_topology_update()
+    def _restore_links(self, link_id):
```
As this method now restores just one link.
```suggestion
    def _restore_link(self, link_id):
```

cmagnobarbosa committed:
Change signature of methods related to restoring administrative state

Apply code review suggestions, renaming methods related to restoring the
administrative state of links and switches, improving readability.

Co-authored-by: Antonio Francisco <antonio@ansp.br>

gleybersonandrade commented on tests/unit/test_main.py:
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
+        self.napp.links_state = {link_id: {"enabled": True}}
+        with self.assertRaises(KeyError):
+            self.napp._restore_link(link_id_fail)
+        self.assertRaises(KeyError)
```
I think this 'assertRaises' can be removed, because it was tested before.

gleybersonandrade commented on tests/unit/test_main.py:
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
+        self.napp.links_state = {link_id_fail: {"enabled": True}}
+        with self.assertRaises(KeyError):
+            self.napp._restore_link(link_id_fail)
+        self.assertRaises(KeyError)
```
Also here

gleybersonandrade commented on tests/unit/test_main.py:
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
+
+        with self.assertRaises(KeyError):
+            self.napp._restore_switch(dpid_fail)
+        self.assertRaises(KeyError)
```
Here

gleybersonandrade commented on tests/unit/test_main.py:
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
+        self.napp.switches_state = {dpid_fail: True}
+        with self.assertRaises(KeyError):
+            self.napp._restore_switch(dpid_fail)
+        self.assertRaises(KeyError)
```
And here

cmagnobarbosa committed:
Refactor the unit tests related to the method `restore_network_status`

Co-authored-by: Gleyberson Andrade <gleybersonandrade@gmail.com>

cmagnobarbosa committed:
Merge branch 'master' into persistence_behavior

cmagnobarbosa committed:
Update openapi.yml and fix linter issues

hdiogenes commented on main.py:
```diff
@@ -92,71 +97,79 @@ def _get_link_from_interface(self, interface):
+            raise KeyError(error)
+        # wait to restore interfaces
+        while not switch.interfaces:
+            time.sleep(self.interface_restore)
```
I think we should have a log.debug here, so we can detect if this code gets stuck for some reason.

hdiogenes commented on settings.py:
```diff
@@ -5,3 +5,7 @@
 # Time (in seconds) to wait before setting a link as up
 LINK_UP_TIMER = 10
+# Time (in seconds) to wait retrieve box from storehouse
+WAIT_PERSISTENCE_BOX_TIMER = 0.1
```
```suggestion
BOX_RESTORE_TIMER = 0.1
```

cmagnobarbosa replied:
Resolved.

hdiogenes commented on storehouse.py:
```diff
@@ -20,15 +25,19 @@ def __init__(self, controller):
-        self.get_stored_box(self.box.box_id)
+        # Wait retrieve or create box in storehouse
+        while not self.box:
+            time.sleep(self.wait_persistence_box)
```
The default here is 0.1 seconds, so a debug may flood the console. Maybe a better option is to add an upper limit: it runs for X times without result, give up and log and error.

cmagnobarbosa replied:
Resolved.

cmagnobarbosa committed:
Add retry limits in waiting conditions

Now, events with waiting conditions have attempts limits. Also, some debug logs
was added. The variables related to waiting conditions have been changed for
improve readability.

Co-authored-by: Humberto Diógenes <hdiogenes@gmail.com>

hdiogenes committed:
[refactoring] Renamed variables, cleared code flow

hdiogenes committed:
Move enable/disable outside try/except

hdiogenes committed:
Created a specific RestoreError exception

Moved code outside try/except

cmagnobarbosa committed:
Fix broken tests and create the exception.py file
